connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/SingleFieldPath.java

```diff
@@ -0,0 +1,585 @@
+import org.apache.kafka.connect.data.Struct;
+import org.apache.kafka.connect.transforms.util.SchemaUtil;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * A FieldPath is composed by one or many field names, known as steps,
+ * to access values within a data object (either {@code Struct} or {@code Map<String, Object>}).a
+ * <p>
+ * If the SMT requires accessing multiple fields on the same data object,
+ * use {@code FieldPaths} instead.
+ * <p>
+ * The field path semantics are defined by the syntax version {@code FieldSyntaxVersion}.
+ * <p>
+ * Paths are calculated once and cached for further access.
+ * <p>
+ * Invariants:
+ * <li>
+ *     <ul>A field path can contain one or more steps</ul>
+ * </li>
+ *
+ * See KIP-821.
+ *
+ * @see FieldSyntaxVersion
+ * @see MultiFieldPaths
+ */
+public class SingleFieldPath implements FieldPath {
+
+    private static final char BACKTICK_CHAR = '`';
+    private static final char DOT_CHAR = '.';
+    private static final char BACKSLASH_CHAR = '\\';
+
+    private static final Cache<String, SingleFieldPath> PATHS_CACHE = new SynchronizedCache<>(new LRUCache<>(16));
+
+    private final String[] path;
+
+    static SingleFieldPath ofV1(String field) {
+        return of(field, FieldSyntaxVersion.V1);
+    }
+
+    static SingleFieldPath ofV2(String field) {
+        return of(field, FieldSyntaxVersion.V2);
+    }
+
+    /**
+     * If version is V2, then paths are cached for further access.
+     *
+     * @param field   field path expression
+     * @param version field syntax version
+     */
+    public static SingleFieldPath of(String field, FieldSyntaxVersion version) {
+        if ((field == null || field.isEmpty()) // empty path
+                || version.equals(FieldSyntaxVersion.V1)) { // or V1
+            return new SingleFieldPath(field, version);
+        } else { // use cache when V2
+            final SingleFieldPath found = PATHS_CACHE.get(field);
+            if (found != null) {
+                return found;
+            } else {
+                final SingleFieldPath fieldPath = new SingleFieldPath(field, version);
+                PATHS_CACHE.put(field, fieldPath);
+                return fieldPath;
+            }
+        }
+    }
+
+    SingleFieldPath(String pathText, FieldSyntaxVersion version) {
+        if (pathText == null || pathText.isEmpty()) { // empty path
+            this.path = new String[] {};
+        } else {
+            switch (version) {
+                case V1: // backward compatibility
+                    this.path = new String[] {pathText};
+                    break;
+                case V2:
+                    path = buildFieldPathV2(pathText);
+                    break;
+                default:
+                    throw new IllegalArgumentException("Unknown syntax version: " + version);
+            }
+        }
+    }
```

Comment:
I think we can simplify this (assuming `buildFieldPathV2` and `escapeBackticks` are made static, and the cache is eliminated) by making the constructor do nothing except take in and store the `path` field, and putting the parsing and version branching logic all in the `of` method. I also don't think we should silently accept the case where `field` is null, but if this is necessary to accommodate an existing use case, let me know and we can probably keep that part.


```suggestion
    public static SingleFieldPath of(String field, FieldSyntaxVersion version) {
        Objects.requireNonNull(field, "Field path cannot be null");
        switch (version) {
            case V1:
                return new SingleFieldPath(new String[] { field } );
            case V2:
                String[] path = buildFieldPathV2(field);
                return new SingleFieldPath(path);
            default:
                throw new IllegalArgumentException("Unknown syntax version");
        }
    }

    SingleFieldPath(String[] path) {
        this.path = path;
    }
```

Reply:
I like it! Adapting the constructor to this approach